pkgs/pkgs-lib/formats.nix

```
@@ -58,9 +59,121 @@ rec {
       jq . "$valuePath"> $out
     '') {};
 
+    /*
+
+    Takes an attrset and a file path and generates a bash snippet that
+    outputs a JSON file at the file path with all secrets in the
+    attrset replaced with the contents of the files they point to in
+    the output JSON.
+
+    When a configuration option accepts an attrset that is finally
+    converted to JSON, this makes it possible to let the user define
+    arbitrary secret values.
+
+    Example:
+      If the file "/path/to/secret" contains the string
+      "topsecretpassword1234",
+
+      genJqSecretsReplacementSnippet
+        {
+          example = [
+            {
+              irrelevant = "not interesting";
+            }
+            {
+              ignored = "ignored attr";
+              relevant = {
+                secret = lib.mkSecret "/path/to/secret";
+              };
+            }
+          ];
+        }
+        "/path/to/output.json"
+
+      would generate a snippet that, when run, outputs the following
+      JSON file at "/path/to/output.json":
+
+      {
+        "example": [
+          {
+            "irrelevant": "not interesting"
+          },
+          {
+            "ignored": "ignored attr",
+            "relevant": {
+              "secret": "topsecretpassword1234"
+            }
+          }
+        ]
+      }   */
+    genJqSecretsReplacementSnippet = set: output:
+      let /*
+        Recurse into a list or an attrset, searching for secrets, and
+        return an attrset where the names are the corresponding jq
+        path where the attrs were found and the values are the values
+        of the attrs.
+
+        Example:
+          getSecretsWithJqPrefix
+            {
+              example = [
+                {
+                  irrelevant = "not interesting";
+                }
+                {
+                  ignored = "ignored attr";
+                  relevant = {
+                    secret = lib.mkSecret "/path/to/secret";
+                  };
+                }
+              ];
+            }
+            -> { ".example[1].relevant.secret" = "/path/to/secret"; }   */
+        getSecretsWithJqPrefix = item:
+          let
+            recurse = prefix: item:
+              if lib.isSecret item then
+                lib.nameValuePair prefix item._secret
+              else if builtins.isAttrs item then
+                lib.mapAttrsToList (name: value: recurse (prefix + "." + name) value) item
+              else if builtins.isList item then
+                lib.imap0 (index: item: recurse (prefix + "[${toString index}]") item) item
+              else
+                [];
+          in lib.listToAttrs (lib.flatten (recurse "" item));
```

Then you also don't need to do the `flatten` here, which imo is a bit ugly since it needs to do a lot of `isList` checks
```suggestion
          in lib.listToAttrs (recurse "" item);
```